pkgs/development/python-modules/pytorch/default.nix

```diff
@@ -72,38 +136,57 @@ in buildPythonPackage rec {
   PYTORCH_BUILD_VERSION = version;
   PYTORCH_BUILD_NUMBER = 0;
 
+  BUILD_NAMEDTENSOR = buildNamedTensor;  # experimental feature
+  USE_SYSTEM_NCCL=true;                  # don't build pytorch's third_party NCCL
+
   # Suppress a weird warning in mkl-dnn, part of ideep in pytorch
   # (upstream seems to have fixed this in the wrong place?)
   # https://github.com/intel/mkl-dnn/commit/8134d346cdb7fe1695a2aa55771071d455fae0bc
-  NIX_CFLAGS_COMPILE = lib.optionals (numpy.blasImplementation == "mkl") [ "-Wno-error=array-bounds" ];
+  # https://github.com/pytorch/pytorch/issues/22346
+  #
+  # Also of interest: pytorch ignores CXXFLAGS uses CFLAGS for both C and C++:
+  # https://github.com/pytorch/pytorch/blob/v1.2.0/setup.py#L17
+  NIX_CFLAGS_COMPILE = lib.optionals (my_numpy.blasImplementation == "mkl") [ "-Wno-error=array-bounds" ];
 
   nativeBuildInputs = [
-     cmake
-     utillinux
-     which
+    cmake
+    utillinux
+    which
+    ninja
   ] ++ lib.optionals cudaSupport [ cudatoolkit_joined ];
 
   buildInputs = [
-     numpy.blas
-  ] ++ lib.optionals cudaSupport [ cudnn ]
+    my_numpy.blas
+  ] ++ lib.optionals cudaSupport [ cudnn my_magma nccl ]
     ++ lib.optionals stdenv.isLinux [ numactl ];
 
   propagatedBuildInputs = [
     cffi
-    numpy
+    my_numpy
     pyyaml
-  ] ++ lib.optional (pythonOlder "3.5") typing;
-
-  checkInputs = [ hypothesis ];
-  checkPhase = ''
-    ${cudaStubEnv}python test/run_test.py --exclude dataloader sparse torch utils thd_distributed distributed cpp_extensions
+  ] ++ lib.optionals openMPISupport [ my_openmpi ]
+    ++ lib.optional (pythonOlder "3.5") typing;
+
+  checkInputs = [ hypothesis ninja ];
+
+  doCheck = false; # tests take a long time for channel release, so doCheck should be overridden only when developing
+  checkPhase = "${cudaStubEnv}python test/run_test.py"
+    + " --exclude utils" # utils requires git, which is not allowed in the check phase
+
+    # Other tests which have been disabled in previous nix derivations of pytorch.
+    # --exclude dataloader sparse torch utils thd_distributed distributed cpp_extensions
+    ;
+  postInstall = ''
+    mkdir $dev
+    cp -r $out/${python.sitePackages}/torch/lib     $dev/lib
+    cp -r $out/${python.sitePackages}/torch/include $dev/include
   '';
```

Comment:
Hi @stites!
I'm running into a new problem on `Darwin` when building hasktorch's `libtorch-ffi`:
```
Preprocessing test suite 'spec' for libtorch-ffi-1.1.0.0..
Building test suite 'spec' for libtorch-ffi-1.1.0.0..
[1 of 5] Compiling BackwardSpec     ( test/BackwardSpec.hs, dist/build/spec/spec-tmp/BackwardSpec.o ) [flags changed]
[2 of 5] Compiling BasicSpec        ( test/BasicSpec.hs, dist/build/spec/spec-tmp/BasicSpec.o ) [flags changed]
[3 of 5] Compiling CudaSpec         ( test/CudaSpec.hs, dist/build/spec/spec-tmp/CudaSpec.o ) [flags changed]
[4 of 5] Compiling MemorySpec       ( test/MemorySpec.hs, dist/build/spec/spec-tmp/MemorySpec.o ) [flags changed]
[5 of 5] Compiling Main             ( test/Spec.hs, dist/build/spec/spec-tmp/Main.o ) [flags changed]
Linking dist/build/spec/spec ...
running tests
Running 1 test suites...
Test suite spec: RUNNING...
dyld: Library not loaded: @rpath/libc10.dylib
  Referenced from: /private/tmp/nix-build-libtorch-ffi-1.1.0.0.drv-0/libtorch-ffi/dist/build/spec/spec
  Reason: image not found
Test suite spec: FAIL
Test suite logged to: dist/test/libtorch-ffi-1.1.0.0-spec.log
0 of 1 test suites (0 of 1 test cases) passed.
builder for '/nix/store/s6av6bgxnsf53n06ryxmc9ksajbwasvi-libtorch-ffi-1.1.0.0.drv' failed with exit code 1
error: build of '/nix/store/s6av6bgxnsf53n06ryxmc9ksajbwasvi-libtorch-ffi-1.1.0.0.drv' failed
```
I tried a bunch of things short of injecting the torch libs via `LD_LIBRARY_PATH`, but `libc10.dylib` remains to be the problem. I suspect it has to do with the fact that macOS has a weird lookup mechanism for this dylibs that needs additional build steps to work properly.
For instance, consider these lines in the MKL derivation: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/science/math/mkl/default.nix#L61-L71
As you can see, `install_name_tool` is applied to every dylib in the package.
See https://stackoverflow.com/questions/33991581/install-name-tool-to-update-a-executable-to-search-for-dylib-in-mac-os-x for more information.